rfqservice/stream.go

```diff
@@ -0,0 +1,284 @@
+	// the peer raw messages subscription.
+	errRecvRawMessages <-chan error
+
+	// incomingMessages is a channel which is populated with incoming
+	// (received) RFQ messages.
+	incomingMessages chan<- rfqmsg.IncomingMsg
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewStreamHandler creates and starts a new RFQ stream handler.
+//
+// TODO(ffranr): Pass in a signer so that we can create a signature over output
+// message fields.
+func NewStreamHandler(ctx context.Context, cfg StreamHandlerCfg,
+	incomingMessages chan<- rfqmsg.IncomingMsg) (*StreamHandler, error) {
+
+	pPorter := cfg.PeerMessagePorter
+	msgChan, peerMsgErrChan, err := pPorter.SubscribeCustomMessages(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to subscribe to custom "+
+			"messages via message transfer handle: %w", err)
+	}
+
+	return &StreamHandler{
+		cfg: cfg,
+
+		recvRawMessages:    msgChan,
+		errRecvRawMessages: peerMsgErrChan,
+
+		incomingMessages: incomingMessages,
+
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// handleIncomingQuoteRequest handles an incoming quote request peer message.
+func (h *StreamHandler) handleIncomingQuoteRequest(
+	wireMsg rfqmsg.WireMessage) error {
+
+	quoteRequest, err := rfqmsg.NewRequestMsgFromWire(wireMsg)
+	if err != nil {
+		return fmt.Errorf("unable to create a quote request message "+
+			"from a wire message: %w", err)
+	}
+
+	// Send the quote request to the RFQ manager.
+	var msg rfqmsg.IncomingMsg = quoteRequest
+	sendSuccess := fn.SendOrQuit(h.incomingMessages, msg, h.Quit)
+	if !sendSuccess {
+		return fmt.Errorf("RFQ stream handler shutting down")
+	}
+
+	return nil
+}
+
+// handleIncomingQuoteAccept handles an incoming quote accept peer message.
+func (h *StreamHandler) handleIncomingQuoteAccept(
+	wireMsg rfqmsg.WireMessage) error {
+
+	quoteAccept, err := rfqmsg.NewAcceptFromWireMsg(wireMsg)
+	if err != nil {
+		return fmt.Errorf("unable to create a quote accept message "+
+			"from a wire message: %w", err)
+	}
+	quoteAccept = quoteAccept
+
+	// TODO(ffranr): At this point, we need to validate that the quote
+	//  accept message is valid. We need to check that the quote accept
+	//  message is associated with a valid quote request. Finally, we need
+	//  to inform the RFQ manager that the quote request has been accepted.
+
+	return nil
+}
+
+// handleIncomingQuoteReject handles an incoming quote reject peer message.
+func (h *StreamHandler) handleIncomingQuoteReject(
+	wireMsg rfqmsg.WireMessage) error {
+
+	quoteReject, err := rfqmsg.NewQuoteRejectFromWireMsg(wireMsg)
+	if err != nil {
+		return fmt.Errorf("unable to create a quote reject message "+
+			"from a wire message: %w", err)
+	}
+	quoteReject = quoteReject
+
+	// TODO(ffranr): At this point, we need to validate that the quote
```

Comment:
Rather than handle it here, if this is meant to be a dumb message passing sub-system, it can delegate the actual handling of this message into a higher layer.

Reply:
Yes, i see what you mean. Maybe with a peer session service or something like that.